File: CS/CalDAVServer.SqlStorage.AspNet/FormsMembershipProvider.cs

using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Linq;
using System.Web;
using System.Web.Configuration;
using System.Web.Security;

namespace CalDAVServer.SqlStorage.AspNet
{
    /// <summary>
    /// Implementation of membership provider which takes user names and passwords from standard section 
    /// in web.config/app.config.
    /// </summary>
    /// <remarks>
    /// It can be replaced with any other existing membership provider in web.config/app.config.
    /// </remarks>
    public class FormsMembershipProvider : MembershipProvider
    {
        public override string ApplicationName
        {
            get
            {
                throw new NotImplementedException();
            }

            set
            {
                throw new NotImplementedException();
            }
        }

        public override bool ChangePassword(string username, string oldPassword, string newPassword)
        {
            throw new NotImplementedException();
        }

        public override bool ChangePasswordQuestionAndAnswer(
            string username,
            string password,
            string newPasswordQuestion,
            string newPasswordAnswer)
        {
            throw new NotImplementedException();
        }

        public override MembershipUser CreateUser(
            string username,
            string password,
            string email,
            string passwordQuestion,
            string passwordAnswer,
            bool isApproved,
            object providerUserKey,
            out MembershipCreateStatus status)
        {
            throw new NotImplementedException();
        }

        public override bool DeleteUser(string username, bool deleteAllRelatedData)
        {
            throw new NotImplementedException();
        }

        public override bool EnablePasswordReset
        {
            get { throw new NotImplementedException(); }
        }

        public override bool EnablePasswordRetrieval
        {
            get { throw new NotImplementedException(); }
        }

        public override MembershipUserCollection FindUsersByEmail(
            string emailToMatch,
            int pageIndex,
            int pageSize,
            out int totalRecords)
        {
            throw new NotImplementedException();
        }

        public override MembershipUserCollection FindUsersByName(
            string usernameToMatch,
            int pageIndex,
            int pageSize,
            out int totalRecords)
        {
            throw new NotImplementedException();
        }

        public override MembershipUserCollection GetAllUsers(int pageIndex, int pageSize, out int totalRecords)
        {
            AuthenticationSection authSection =
                (AuthenticationSection)WebConfigurationManager.GetWebApplicationSection("system.web/authentication");

            NameValueCollection emailsSection =
                (NameValueCollection)System.Configuration.ConfigurationManager.GetSection("emails");

            totalRecords = authSection.Forms.Credentials.Users.Count;
            MembershipUserCollection users = new MembershipUserCollection();


            for (int i = pageIndex * pageSize;
                i < Math.Min(totalRecords, pageIndex * pageSize + pageSize);
                i++)
            {
                FormsAuthenticationUser user = authSection.Forms.Credentials.Users[i];
                string email = null;
                email = emailsSection[user.Name];

                users.Add(new MembershipUser(
                        "FormsProvider",
                        user.Name,
                        null,
                        email,
                        null,
                        null,
                        true,
                        false,
                        // do not use DateTime.MinValue because some WebDAV clients may not properly parse it.
                        new DateTime(2000, 1, 1),
                        new DateTime(2000, 1, 1),
                        new DateTime(2000, 1, 1),
                        new DateTime(2000, 1, 1),
                        new DateTime(2000, 1, 1))
                );
            }

            return users;
        }

        public override int GetNumberOfUsersOnline()
        {
            throw new NotImplementedException();
        }

        public override string GetPassword(string username, string answer)
        {
            throw new NotImplementedException();
        }

        public override MembershipUser GetUser(string username, bool userIsOnline)
        {
            string email = null;

            AuthenticationSection authSection =
                (AuthenticationSection)WebConfigurationManager.GetWebApplicationSection("system.web/authentication");
            FormsAuthenticationUser user = authSection.Forms.Credentials.Users[username.ToLower()];

            if (user != null)
            {
                NameValueCollection emailsSection =
                    (NameValueCollection)System.Configuration.ConfigurationManager.GetSection("emails");
                email = emailsSection[user.Name];
                return new MembershipUser(
                    "FormsProvider",
                    username,
                    null,
                    email,
                    null,
                    null,
                    true,
                    false,
                    // do not use DateTime.MinValue because some WebDAV clients may not properly parse it.
                    new DateTime(2000, 1, 1),
                    new DateTime(2000, 1, 1),
                    new DateTime(2000, 1, 1),
                    new DateTime(2000, 1, 1),
                    new DateTime(2000, 1, 1));
            }

            return null;
        }

        public override MembershipUser GetUser(object providerUserKey, bool userIsOnline)
        {
            throw new NotImplementedException();
        }

        public override string GetUserNameByEmail(string email)
        {
            throw new NotImplementedException();
        }

        public override int MaxInvalidPasswordAttempts
        {
            get { throw new NotImplementedException(); }
        }

        public override int MinRequiredNonAlphanumericCharacters
        {
            get { throw new NotImplementedException(); }
        }

        public override int MinRequiredPasswordLength
        {
            get { throw new NotImplementedException(); }
        }

        public override int PasswordAttemptWindow
        {
            get { throw new NotImplementedException(); }
        }

        public override MembershipPasswordFormat PasswordFormat
        {
            get { return MembershipPasswordFormat.Clear; }
        }

        public override string PasswordStrengthRegularExpression
        {
            get { throw new NotImplementedException(); }
        }

        public override bool RequiresQuestionAndAnswer
        {
            get { throw new NotImplementedException(); }
        }

        public override bool RequiresUniqueEmail
        {
            get { throw new NotImplementedException(); }
        }

        public override string ResetPassword(string username, string answer)
        {
            throw new NotImplementedException();
        }

        public override bool UnlockUser(string userName)
        {
            throw new NotImplementedException();
        }

        public override void UpdateUser(MembershipUser user)
        {
            throw new NotImplementedException();
        }

        public override bool ValidateUser(string username, string password)
        {
#pragma warning disable 0618
            return FormsAuthentication.Authenticate(username, password);
#pragma warning restore 0618
        }
    }
}

File: VB/CalDAVServer.SqlStorage.AspNet/FormsMembershipProvider.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.Specialized
Imports System.Linq
Imports System.Web
Imports System.Web.Configuration
Imports System.Web.Security

''' <summary>
''' Implementation of membership provider which takes user names and passwords from standard section 
''' in web.config/app.config.
''' </summary>
''' <remarks>
''' It can be replaced with any other existing membership provider in web.config/app.config.
''' </remarks>
Public Class FormsMembershipProvider
    Inherits MembershipProvider

    Public Overrides Property ApplicationName As String
        Get
            Throw New NotImplementedException()
        End Get

        Set(ByVal value As String)
            Throw New NotImplementedException()
        End Set
    End Property

    Public Overrides Function ChangePassword(username As String, oldPassword As String, newPassword As String) As Boolean
        Throw New NotImplementedException()
    End Function

    Public Overrides Function ChangePasswordQuestionAndAnswer(username As String,
                                                             password As String,
                                                             newPasswordQuestion As String,
                                                             newPasswordAnswer As String) As Boolean
        Throw New NotImplementedException()
    End Function

    Public Overrides Function CreateUser(username As String,
                                        password As String,
                                        email As String,
                                        passwordQuestion As String,
                                        passwordAnswer As String,
                                        isApproved As Boolean,
                                        providerUserKey As Object,
                                        ByRef status As MembershipCreateStatus) As MembershipUser
        Throw New NotImplementedException()
    End Function

    Public Overrides Function DeleteUser(username As String, deleteAllRelatedData As Boolean) As Boolean
        Throw New NotImplementedException()
    End Function

    Public Overrides ReadOnly Property EnablePasswordReset As Boolean
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides ReadOnly Property EnablePasswordRetrieval As Boolean
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides Function FindUsersByEmail(emailToMatch As String,
                                              pageIndex As Integer,
                                              pageSize As Integer,
                                              ByRef totalRecords As Integer) As MembershipUserCollection
        Throw New NotImplementedException()
    End Function

    Public Overrides Function FindUsersByName(usernameToMatch As String,
                                             pageIndex As Integer,
                                             pageSize As Integer,
                                             ByRef totalRecords As Integer) As MembershipUserCollection
        Throw New NotImplementedException()
    End Function

    Public Overrides Function GetAllUsers(pageIndex As Integer, pageSize As Integer, ByRef totalRecords As Integer) As MembershipUserCollection
        Dim authSection As AuthenticationSection = CType(WebConfigurationManager.GetWebApplicationSection("system.web/authentication"), AuthenticationSection)
        Dim emailsSection As NameValueCollection = CType(System.Configuration.ConfigurationManager.GetSection("emails"), NameValueCollection)
        totalRecords = authSection.Forms.Credentials.Users.Count
        Dim users As MembershipUserCollection = New MembershipUserCollection()
        For i As Integer = pageIndex * pageSize To Math.Min(totalRecords, pageIndex * pageSize + pageSize) - 1
            Dim user As FormsAuthenticationUser = authSection.Forms.Credentials.Users(i)
            Dim email As String = Nothing
            email = emailsSection(user.Name)
            users.Add(New MembershipUser("FormsProvider",
                                        user.Name,
                                        Nothing,
                                        email,
                                        Nothing,
                                        Nothing,
                                        True,
                                        False,
                                        New DateTime(2000, 1, 1),
                                        New DateTime(2000, 1, 1),
                                        New DateTime(2000, 1, 1),
                                        New DateTime(2000, 1, 1),
                                        New DateTime(2000, 1, 1)))
        Next

        Return users
    End Function

    Public Overrides Function GetNumberOfUsersOnline() As Integer
        Throw New NotImplementedException()
    End Function

    Public Overrides Function GetPassword(username As String, answer As String) As String
        Throw New NotImplementedException()
    End Function

    Public Overrides Function GetUser(username As String, userIsOnline As Boolean) As MembershipUser
        Dim email As String = Nothing
        Dim authSection As AuthenticationSection = CType(WebConfigurationManager.GetWebApplicationSection("system.web/authentication"), AuthenticationSection)
        Dim user As FormsAuthenticationUser = authSection.Forms.Credentials.Users(username.ToLower())
        If user IsNot Nothing Then
            Dim emailsSection As NameValueCollection = CType(System.Configuration.ConfigurationManager.GetSection("emails"), NameValueCollection)
            email = emailsSection(user.Name)
            Return New MembershipUser("FormsProvider",
                                     username,
                                     Nothing,
                                     email,
                                     Nothing,
                                     Nothing,
                                     True,
                                     False,
                                     New DateTime(2000, 1, 1),
                                     New DateTime(2000, 1, 1),
                                     New DateTime(2000, 1, 1),
                                     New DateTime(2000, 1, 1),
                                     New DateTime(2000, 1, 1))
        End If

        Return Nothing
    End Function

    Public Overrides Function GetUser(providerUserKey As Object, userIsOnline As Boolean) As MembershipUser
        Throw New NotImplementedException()
    End Function

    Public Overrides Function GetUserNameByEmail(email As String) As String
        Throw New NotImplementedException()
    End Function

    Public Overrides ReadOnly Property MaxInvalidPasswordAttempts As Integer
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides ReadOnly Property MinRequiredNonAlphanumericCharacters As Integer
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides ReadOnly Property MinRequiredPasswordLength As Integer
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides ReadOnly Property PasswordAttemptWindow As Integer
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides ReadOnly Property PasswordFormat As MembershipPasswordFormat
        Get
            Return MembershipPasswordFormat.Clear
        End Get
    End Property

    Public Overrides ReadOnly Property PasswordStrengthRegularExpression As String
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides ReadOnly Property RequiresQuestionAndAnswer As Boolean
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides ReadOnly Property RequiresUniqueEmail As Boolean
        Get
            Throw New NotImplementedException()
        End Get
    End Property

    Public Overrides Function ResetPassword(username As String, answer As String) As String
        Throw New NotImplementedException()
    End Function

    Public Overrides Function UnlockUser(userName As String) As Boolean
        Throw New NotImplementedException()
    End Function

    Public Overrides Sub UpdateUser(user As MembershipUser)
        Throw New NotImplementedException()
    End Sub

    Public Overrides Function ValidateUser(username As String, password As String) As Boolean
        Return FormsAuthentication.Authenticate(username, password)
    End Function
End Class
